barretenberg/cpp/src/barretenberg/ultra_honk/MERGE_PROTOCOL.md

```diff
@@ -0,0 +1,607 @@
+│ │     │non-ops  │         │   ops    │  ops    │non-ops   │      │
+│ └────┴──────────┴──────────┴──────────┴──────────┴──────────┘      │
+│  ↑              ↑                                 ↑                 │
+│  For            Left table in                     Right table in    │
+│  shiftability   final merge                       final merge       │
+│                 (degree check)                    (ecc_op_wires)    │
+│                                                                     │
+│ ZK Masking Strategy:                                                │
+│ • 3 random non-ops at START (from tail kernel, prepended)           │
+│ • 2 random non-ops at END (from hiding kernel, appended)            │
+│ • 1 valid random ECC op (added during tail verification)            │
+│                                                                     │
+│ → Real application operations are SURROUNDED by randomness          │
+└─────────────────────────────────────────────────────────────────────┘
+                              ↓
+┌─────────────────────────────────────────────────────────────────────┐
+│ Final Proof Generation (3 Protocols)                               │
+├─────────────────────────────────────────────────────────────────────┤
+│                                                                     │
+│ 1. MERGE PROTOCOL (THIS PROTOCOL)                                  │
+│    Final merge (APPEND mode):                                      │
+│    • L = T_prev (all previous ops including tail with 3 random)    │
+│    • R = t_hiding (hiding kernel with 2 random non-ops at end)     │
+│    • Proves: M = L + X^ℓ · R and deg(L) < ℓ                        │
+│    • Commitments: [M_j], [G] (new); [L_j], [R_j] (from transcript) │
+│                                                                     │
+│ 2. TRANSLATOR PROTOCOL                                             │
+│    • Proves BN254 ↔ Grumpkin translation correctness              │
+│    • Uses SAME commitments [M_j] as Merge (copy-constrained)       │
+│    • All 6 random ops contribute to ZK                             │
+│                                                                     │
+│ 3. ECCVM PROTOCOL                                                   │
+│    • Proves ECC operations executed correctly                      │
+│    • Only real ops + 1 valid random op (excludes 5 random non-ops) │
+│                                                                     │
+│ → Combined into complete CHONK proof for recursive verification    │
+└─────────────────────────────────────────────────────────────────────┘
+```
+
+**Key Insights:**
+
+1. **Incremental Merge**: Merge verification happens **at each circuit accumulation**
+2. **Mode Switch**: Most circuits use PREPEND mode; **hiding kernel uses APPEND mode**
+3. **Randomness Placement**: The usage of PREPEND (tail) and APPEND (hiding) modes ensures random ops **surround** real application ops
```

Comment:
not sure "surround(ing) real application ops" is revelant haha but maybe the relevant insight here is that they';re placed in such a way that they're easy to handle downstream in the translator